File: tools/c4_phase2a_runtime_spike/vm.mjs
// tools/c4_phase2a_runtime_spike/vm.mjs — the JS dispatch loop.
//
// ⛔ MEASUREMENT INSTRUMENT. See isa.md. Not a Pine runtime.

import fs from 'node:fs'

const PUSH_CONST = 0, PUSH_SERIES = 1, PUSH_HIST = 2,
  ADD = 3, SUB = 4, MUL = 5, DIV = 6, LT = 7,
  LOAD_LOCAL = 8, STORE_LOCAL = 9, LOAD_PERSIST = 10, STORE_PERSIST = 11,
  JUMP_IF_FALSE = 12, JUMP = 13,
  ARR_PUSH = 14, ARR_GET = 15, ARR_SIZE = 16,
  EMIT = 17, POP = 18, HALT = 19

export function run(program, series, bars, limits) {
  const code = program.code
  const consts = program.consts
  const stack = new Float64Array(64)
  const locals = new Float64Array(program.locals)
  const persist = new Float64Array(program.persists)
  const arrays = []
  for (let i = 0; i < program.arrays; i += 1) arrays.push([])
  const out = new Float64Array(bars)
  const maxSteps = limits && limits.maxSteps ? limits.maxSteps : Infinity
  let steps = 0

  for (let bar = 0; bar < bars; bar += 1) {
    let sp = 0
    let pc = 0
    for (;;) {
      const op = code[pc * 3]
      const a = code[pc * 3 + 1]
      const b = code[pc * 3 + 2]
      pc += 1
      steps += 1
      if (steps > maxSteps) throw new Error('INSTRUCTION_LIMIT_EXCEEDED')
      switch (op) {
        case PUSH_CONST: stack[sp++] = consts[a]; break
        case PUSH_SERIES: stack[sp++] = series[a][bar]; break
        case PUSH_HIST: {
          const off = b === -1 ? stack[--sp] : b
          const idx = bar - off
          stack[sp++] = (idx >= 0 && idx <= bar) ? series[a][idx] : NaN
          break
        }
        case ADD: { const y = stack[--sp], x = stack[--sp]; stack[sp++] = (x !== x || y !== y) ? NaN : x + y; break }
        case SUB: { const y = stack[--sp], x = stack[--sp]; stack[sp++] = (x !== x || y !== y) ? NaN : x - y; break }
        case MUL: { const y = stack[--sp], x = stack[--sp]; stack[sp++] = (x !== x || y !== y) ? NaN : x * y; break }
        case DIV: { const y = stack[--sp], x = stack[--sp]; stack[sp++] = (x !== x || y !== y || y === 0) ? NaN : x / y; break }
        case LT: { const y = stack[--sp], x = stack[--sp]; stack[sp++] = (x !== x || y !== y) ? NaN : (x < y ? 1 : 0); break }
        case LOAD_LOCAL: stack[sp++] = locals[a]; break
        case STORE_LOCAL: locals[a] = stack[--sp]; break
        case LOAD_PERSIST: stack[sp++] = persist[a]; break
        case STORE_PERSIST: persist[a] = stack[--sp]; break
        case JUMP_IF_FALSE: { const t = stack[--sp]; if (t !== t || t === 0) pc = a; break }
        case JUMP: pc = a; break
        case ARR_PUSH: arrays[a].push(stack[--sp]); break
        case ARR_GET: { const i = stack[--sp] | 0; const arr = arrays[a]; stack[sp++] = (i >= 0 && i < arr.length) ? arr[i] : NaN; break }
        case ARR_SIZE: stack[sp++] = arrays[a].length; break
        case EMIT: out[bar] = stack[--sp]; break
        case POP: sp -= 1; break
        case HALT: break
        default: throw new Error('bad opcode ' + op)
      }
      if (op === HALT) break
    }
  }
  return { out, steps }
}

export function synthBars(n) {
  const close = new Float64Array(n)
  const open = new Float64Array(n)
  const high = new Float64Array(n)
  const low = new Float64Array(n)
  let p = 100
  for (let i = 0; i < n; i += 1) {
    p += Math.sin(i / 7) * 0.4 + Math.cos(i / 13) * 0.2
    close[i] = p; open[i] = p - 0.1; high[i] = p + 0.5; low[i] = p - 0.5
  }
  return [open, high, low, close]
}

// ── CLI ─────────────────────────────────────────────────────────────────────
const isMain = process.argv[1] && process.argv[1].endsWith('vm.mjs')
if (isMain) {
  const args = process.argv.slice(2)
  const arg = (f, d) => { const i = args.indexOf(f); return i >= 0 && args[i + 1] ? args[i + 1] : d }
  const program = JSON.parse(fs.readFileSync(arg('--program', 'program.json'), 'utf8'))
  const bars = Number(arg('--bars', 300))
  const reps = Number(arg('--reps', 5))
  const symbols = Number(arg('--symbols', 1))
  const series = synthBars(bars)

  const t0 = performance.now()
  let checksum = 0
  let steps = 0
  for (let s = 0; s < symbols; s += 1) {
    const r = run(program, series, bars, {})
    steps = r.steps
    checksum += r.out[bars - 1]
  }
  const first = performance.now() - t0

  const times = []
  for (let r = 0; r < reps; r += 1) {
    const a = performance.now()
    for (let s = 0; s < symbols; s += 1) run(program, series, bars, {})
    times.push(performance.now() - a)
  }
  times.sort((x, y) => x - y)
  const r0 = run(program, series, bars, {})
  process.stdout.write(JSON.stringify({
    host: 'node ' + process.version,
    bars, symbols, steps,
    firstRunMs: first,
    medianMs: times[Math.floor(times.length / 2)],
    minMs: times[0],
    checksum,
    head: Array.from(r0.out.slice(0, 5)),
    tail: Array.from(r0.out.slice(-3)),
  }))
}

File: tools/c4_phase2a_runtime_spike/vm.py
"""tools/c4_phase2a_runtime_spike/vm.py — the Python dispatch loop.

⛔ MEASUREMENT INSTRUMENT. See isa.md. Not a Pine runtime.

⭐ WRITTEN THE WAY A PYTHON VM WOULD ACTUALLY BE WRITTEN IF IT HAD TO BE FAST:
local-variable binding of every hot name, a list-based stack with explicit index,
`math.isnan` avoided in favour of `x != x`, and no attribute lookups inside the
dispatch loop. Measuring a naive implementation would understate Python and make
the architecture decision on a strawman.
"""

import json
import sys
import time

PUSH_CONST, PUSH_SERIES, PUSH_HIST = 0, 1, 2
ADD, SUB, MUL, DIV, LT = 3, 4, 5, 6, 7
LOAD_LOCAL, STORE_LOCAL, LOAD_PERSIST, STORE_PERSIST = 8, 9, 10, 11
JUMP_IF_FALSE, JUMP = 12, 13
ARR_PUSH, ARR_GET, ARR_SIZE = 14, 15, 16
EMIT, POP, HALT = 17, 18, 19

NAN = float("nan")


def run(program, series, bars, max_steps=None):
    code = program["code"]
    consts = program["consts"]
    stack = [0.0] * 64
    locals_ = [0.0] * program["locals"]
    persist = [0.0] * program["persists"]
    arrays = [[] for _ in range(program["arrays"])]
    out = [0.0] * bars
    steps = 0
    limit = max_steps if max_steps is not None else float("inf")

    for bar in range(bars):
        sp = 0
        pc = 0
        while True:
            base = pc * 3
            op = code[base]
            a = code[base + 1]
            b = code[base + 2]
            pc += 1
            steps += 1
            if steps > limit:
                raise RuntimeError("INSTRUCTION_LIMIT_EXCEEDED")

            if op == PUSH_CONST:
                stack[sp] = consts[a]; sp += 1
            elif op == PUSH_SERIES:
                stack[sp] = series[a][bar]; sp += 1
            elif op == PUSH_HIST:
                if b == -1:
                    sp -= 1
                    off = stack[sp]
                else:
                    off = b
                idx = bar - int(off)
                stack[sp] = series[a][idx] if 0 <= idx <= bar else NAN
                sp += 1
            elif op == ADD:
                sp -= 1; y = stack[sp]; sp -= 1; x = stack[sp]
                stack[sp] = NAN if (x != x or y != y) else x + y; sp += 1
            elif op == SUB:
                sp -= 1; y = stack[sp]; sp -= 1; x = stack[sp]
                stack[sp] = NAN if (x != x or y != y) else x - y; sp += 1
            elif op == MUL:
                sp -= 1; y = stack[sp]; sp -= 1; x = stack[sp]
                stack[sp] = NAN if (x != x or y != y) else x * y; sp += 1
            elif op == DIV:
                sp -= 1; y = stack[sp]; sp -= 1; x = stack[sp]
                stack[sp] = NAN if (x != x or y != y or y == 0) else x / y; sp += 1
            elif op == LT:
                sp -= 1; y = stack[sp]; sp -= 1; x = stack[sp]
                stack[sp] = NAN if (x != x or y != y) else (1.0 if x < y else 0.0); sp += 1
            elif op == LOAD_LOCAL:
                stack[sp] = locals_[a]; sp += 1
            elif op == STORE_LOCAL:
                sp -= 1; locals_[a] = stack[sp]
            elif op == LOAD_PERSIST:
                stack[sp] = persist[a]; sp += 1
            elif op == STORE_PERSIST:
                sp -= 1; persist[a] = stack[sp]
            elif op == JUMP_IF_FALSE:
                sp -= 1; t = stack[sp]
                if t != t or t == 0:
                    pc = a
            elif op == JUMP:
                pc = a
            elif op == ARR_PUSH:
                sp -= 1; arrays[a].append(stack[sp])
            elif op == ARR_GET:
                sp -= 1; i = int(stack[sp]); arr = arrays[a]
                stack[sp] = arr[i] if 0 <= i < len(arr) else NAN; sp += 1
            elif op == ARR_SIZE:
                stack[sp] = float(len(arrays[a])); sp += 1
            elif op == EMIT:
                sp -= 1; out[bar] = stack[sp]
            elif op == POP:
                sp -= 1
            elif op == HALT:
                break
            else:
                raise RuntimeError("bad opcode %r" % op)
    return out, steps


def synth_bars(n):
    import math
    close = [0.0] * n
    openp = [0.0] * n
    high = [0.0] * n
    low = [0.0] * n
    p = 100.0
    for i in range(n):
        p += math.sin(i / 7) * 0.4 + math.cos(i / 13) * 0.2
        close[i] = p
        openp[i] = p - 0.1
        high[i] = p + 0.5
        low[i] = p - 0.5
    return [openp, high, low, close]


def _arg(flag, default):
    if flag in sys.argv:
        return sys.argv[sys.argv.index(flag) + 1]
    return default


if __name__ == "__main__":
    with open(_arg("--program", "program.json"), encoding="utf-8") as fh:
        program = json.load(fh)
    bars = int(_arg("--bars", 300))
    reps = int(_arg("--reps", 5))
    symbols = int(_arg("--symbols", 1))
    series = synth_bars(bars)

    t0 = time.perf_counter()
    checksum = 0.0
    steps = 0
    for _ in range(symbols):
        out, steps = run(program, series, bars)
        checksum += out[-1]
    first = (time.perf_counter() - t0) * 1000.0

    times = []
    for _ in range(reps):
        a = time.perf_counter()
        for _ in range(symbols):
            run(program, series, bars)
        times.append((time.perf_counter() - a) * 1000.0)
    times.sort()
    out, _ = run(program, series, bars)
    print(json.dumps({
        "host": "python " + sys.version.split()[0],
        "bars": bars, "symbols": symbols, "steps": steps,
        "firstRunMs": first,
        "medianMs": times[len(times) // 2],
        "minMs": times[0],
        "checksum": checksum,
        "head": out[:5],
        "tail": out[-3:],
    }))
